### PR TITLE
bpo-4963: Fix for initialization and non-deterministic behavior issues in mimetypes

### PR DESCRIPTION
**Pull request by davidkhess**
Tracked in:

https://bugs.python.org/issue4963

In particular, here's my thoughts on it which led to this PR:

https://bugs.python.org/msg293626

https://bugs.python.org/issue4963

### CODE REVIEW

File: Doc/library/mimetypes.rst
Comment (brettcannon) [inline Doc/library/mimetypes.rst]: Since `files` is already set to `None` you don't have to explicitly call out what happens if you don't specify an argument.

**Comment by davidkhess** [comment]
I see a new check. Should I write a news entry for this? Or will the committer?

**Comment by brettcannon** [comment]
@davidkhess It's easier for us if you create the news entry because if a core dev decides to merge this they can edit the file easily through GitHub's web UI while having to add a file is much more burdensome.

**Comment by davidkhess** [comment]
News entry added.

**Comment by davidkhess** [comment]
@bitdancer, anything I can do to help this along?

**Comment by davidkhess** [comment]
@bitdancer, anything I can do to help this along?

**Comment by davidkhess** [comment]
@brettcannon @bitdancer, anything I can do to help this along?

I'm willing to resolve the conflict above but I don't want to waste my time if this PR is never going to be merged.

**Comment by brettcannon** [comment]
Unfortunately I don't know this module well enough to feel comfortable reviewing it. I would try pinging the issue to ask if someone could review it.

**Comment by davidkhess** [comment]
Pinged.

File: Lib/mimetypes.py
Comment (serhiy-storchaka) [inline Lib/mimetypes.py]: This no longer needed in 3.7+.

Comment (davidkhess) [reply Lib/mimetypes.py]: Should I remove all usage of OrderedDict from the PR and let others add it back for back porting it to 2.7? Or should I do something like set OrderedDict = dict if Python version >= 3.7?

**Review by tiran — DISMISSED** [review]
Please resolve the merge conflict

File: Lib/mimetypes.py
Comment (zooba) [reply Lib/mimetypes.py]: The order of the items in the list is a critical part of the fix - I'd vote for retaining the explicit use of OrderedDict.

**Review by zooba — APPROVED** [review]

**Comment by zooba** [comment]
The merge conflict does need to be fixed, and it looks like it'll be a bit of work to figure out exactly what needs merging.

File: Lib/mimetypes.py
Comment (serhiy-storchaka) [reply Lib/mimetypes.py]: Adding OrderedDict makes this PR more complex, and the resulting code clumsy, and this is absolutely unneeded in 3.7+. I think it would be better to merge a simple fix into master and 3.7, and prepare more complex backport for 3.6.

**Comment by davidkhess** [comment]
I'll happily fix the merge conflict but I could use some clear direction on OrderedDict or not?

Personally I vote for explicit with OrderedDict.

**Review by auvipy — CHANGES_REQUESTED** [review]
pease rebase.

File: Lib/mimetypes.py
Comment (csabella) [reply Lib/mimetypes.py]: Since 3.6 is only receiving security fixes, there's no need to worry about the backport to 3.6 anymore.  So, you can remove the OrderedDict as this will only be in 3.7 and 3.8.

**Comment by csabella** [comment]
@davidkhess, you can remove the OrderedDict as this point since this won't be applied to 3.6.  Since this was close to being merged before, pending the discussion of the OrderedDict, if you fix the merge conflict and remove the OrderedDict, I think this will be able to be merged.  Thank you for your contribution and for your patience.   🙂

**Comment by auvipy** [comment]
@davidkhess If you allow me to push on your branch I can help to get this merged

**Comment by davidkhess** [comment]
I'm making the requested changes now.

**Comment by davidkhess** [comment]
@auvipy @csabella good to go.

**Comment by csabella** [comment]
@serhiy-storchaka and @zooba, the merge conflicts have been resolved.  I know this was approved last year, but I wanted to make sure you were still OK with it before merging.  Thanks!

**Review by auvipy — APPROVED** [review]

**Comment by adeadman** [comment]
This change has resulted in a difference in behaviour from previous versions of Python in that even when specifying `files` as a parameter to `init()`, the known files are loaded, which is not the case in previous versions. This causes problems when an application wishes to use the mimetypes database with an internal-only mapping of mimetypes, to work around inconsistencies with mimetypes in the known files list (which can vary between systems).

Unfortunately this change wasn't documented in the changelog either, nor in the docstring comments, so it can only be found by examining the commit history for the code in cpython.

Would it be appropriate to add another kwarg to `init()` indicating that knownfiles are to be ignored/excluded from the DB initialization?

**Comment by pombredanne** [comment]
@adeadman posted a message inline as a reply to mine at https://github.com/python/cpython/commit/9fc720e5e4f772598013ea48a3f0d22b2b6b04fa#r45794801

In essence this fixed was supposed to bring in a deterministic behaviour and actually had the complete opposite effect. I still consider this a regression. I stopped using mimetypes for now as it is useless as it is now.

**Comment by tiran** [comment]
Could you please open a new bug report? Comments on closed PRs tend to get lost.

**Pull request merged**
